general/device.hpp

```diff
@@ -102,19 +109,27 @@ class Device
        is allowed. */
    unsigned long allowed_backends;
 
+   MemoryType mem_type;    ///< Current Device MemoryType
+   MemoryClass mem_class;  ///< Current Device MemoryClass
+
    Device()
       : mode(Device::SEQUENTIAL),
         backends(Backend::CPU),
-        allowed_backends(backends) { }
+        allowed_backends(backends),
+        mem_type(MemoryType::HOST),
+        mem_class(MemoryClass::HOST)
+   { }
    Device(Device const&);
    void operator=(Device const&);
-   static Device& Get() { static Device singleton; return singleton; }
+   static Device& Get() { return device_singleton; }
```

Comment:
This looks better to me too, but just note that the initialization time is different (probably not a problem here).

Reply:
This was actually causing the initialization code for `singleton` to be inlined everywhere which was not obvious to me. So the moral of the story is: never define `static` variables inside functions that you want inlined.